In [1]:
import pandas as pd
import json 

In [2]:
import os

#to store my json files
os.makedirs("json_files2", exist_ok=True)

# Load and explore data 

In [3]:
df = pd.read_csv("Sample - Superstore.csv", sep=",", encoding="cp1252")
df.head(10)

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164
5,6,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,FUR-FU-10001487,Furniture,Furnishings,Eldon Expressions Wood and Plastic Desk Access...,48.8600,7,0.00,14.1694
6,7,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-AR-10002833,Office Supplies,Art,Newell 322,7.2800,4,0.00,1.9656
7,8,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,TEC-PH-10002275,Technology,Phones,Mitel 5320 IP Phone VoIP phone,907.1520,6,0.20,90.7152
8,9,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-BI-10003910,Office Supplies,Binders,DXL Angle-View Binders with Locking Rings by S...,18.5040,3,0.20,5.7825
9,10,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-AP-10002892,Office Supplies,Appliances,Belkin F5C206VTEL 6 Outlet Surge,114.9000,5,0.00,34.4700


In [4]:
#check the shape of the data. 9994 rows and 21 columns.
print(df.shape)

(9994, 21)


In [5]:
#check the columns contained in the data
print(df.columns)

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State',
       'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category',
       'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit'],
      dtype='object')


In [6]:
#check how many null entries per column. The data does not contain any null values and it is therefore clean
print(df.isna().sum())

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64


In [7]:
#get the datatypes of each columns 
print(df.dtypes)

Row ID             int64
Order ID          object
Order Date        object
Ship Date         object
Ship Mode         object
Customer ID       object
Customer Name     object
Segment           object
Country           object
City              object
State             object
Postal Code        int64
Region            object
Product ID        object
Category          object
Sub-Category      object
Product Name      object
Sales            float64
Quantity           int64
Discount         float64
Profit           float64
dtype: object


In [8]:
#get some statistical information about the numerical columns 
print(df.describe())

            Row ID   Postal Code         Sales     Quantity     Discount  \
count  9994.000000   9994.000000   9994.000000  9994.000000  9994.000000   
mean   4997.500000  55190.379428    229.858001     3.789574     0.156203   
std    2885.163629  32063.693350    623.245101     2.225110     0.206452   
min       1.000000   1040.000000      0.444000     1.000000     0.000000   
25%    2499.250000  23223.000000     17.280000     2.000000     0.000000   
50%    4997.500000  56430.500000     54.490000     3.000000     0.200000   
75%    7495.750000  90008.000000    209.940000     5.000000     0.200000   
max    9994.000000  99301.000000  22638.480000    14.000000     0.800000   

            Profit  
count  9994.000000  
mean     28.656896  
std     234.260108  
min   -6599.978000  
25%       1.728750  
50%       8.666500  
75%      29.364000  
max    8399.976000  


In [9]:
#get the number of unique values for the categorical columns 
print("For the Order ID column: ", df["Order ID"].nunique())
print("For the Order Date column: ", df["Order Date"].nunique())
print("For the Ship Date column: ", df["Ship Date"].nunique())
print("For the Ship Mode column: ", df["Ship Mode"].nunique())
print("For the Customer ID column: ", df["Customer ID"].nunique())
print("For the Customer Name column: ", df["Customer Name"].nunique())
print("For the Segment column: ", df["Segment"].nunique())
print("For the Country column: ", df["Country"].nunique())
print("For the State column: ", df["State"].nunique())
print("For the Postal Code column: ", df["Postal Code"].nunique())
print("For the Region column: ", df["Region"].nunique())
print("For the Product ID column: ", df["Product ID"].nunique())
print("For the Category column: ", df["Category"].nunique())
print("For the Sub-Category column: ", df["Sub-Category"].nunique())
print("For the Product Name column: ", df["Product Name"].nunique())



For the Order ID column:  5009
For the Order Date column:  1237
For the Ship Date column:  1334
For the Ship Mode column:  4
For the Customer ID column:  793
For the Customer Name column:  793
For the Segment column:  3
For the Country column:  1
For the State column:  49
For the Postal Code column:  631
For the Region column:  4
For the Product ID column:  1862
For the Category column:  3
For the Sub-Category column:  17
For the Product Name column:  1850


In [10]:
#explore the unique values for the columns that have a small number of unique values 
print("The Ship Mode column contains: ", df["Ship Mode"].unique())
print("The Segment column contains: ", df["Segment"].unique())
print("The Country column contains: ", df["Country"].unique())
print("The Region column contains: ", df["Region"].unique())
print("For the Category column contains: ", df["Category"].unique())

The Ship Mode column contains:  ['Second Class' 'Standard Class' 'First Class' 'Same Day']
The Segment column contains:  ['Consumer' 'Corporate' 'Home Office']
The Country column contains:  ['United States']
The Region column contains:  ['South' 'West' 'Central' 'East']
For the Category column contains:  ['Furniture' 'Office Supplies' 'Technology']


In [11]:
#It can be noticed that the dates are of type object. They need to be converted to a datetime object 
df["Order Date"] = pd.to_datetime(df["Order Date"])
df["Ship Date"] = pd.to_datetime(df["Ship Date"])


In [12]:
#check if it was well converted 
print(df.dtypes)

Row ID                    int64
Order ID                 object
Order Date       datetime64[ns]
Ship Date        datetime64[ns]
Ship Mode                object
Customer ID              object
Customer Name            object
Segment                  object
Country                  object
City                     object
State                    object
Postal Code               int64
Region                   object
Product ID               object
Category                 object
Sub-Category             object
Product Name             object
Sales                   float64
Quantity                  int64
Discount                float64
Profit                  float64
dtype: object


In [13]:
#check the data range for these two columns 
range_order = str(df["Order Date"].dt.date.min()) + " to " + str(df["Order Date"].dt.date.max())
print("For the Order Date: ", range_order)
range_ship = str(df["Ship Date"].dt.date.min()) + " to " + str(df["Ship Date"].dt.date.max())
print("For the Ship Date: ", range_ship)

For the Order Date:  2014-01-03 to 2017-12-30
For the Ship Date:  2014-01-07 to 2018-01-05


In [14]:
#explore a little further the profit column 
print(df["Profit"].describe())

count    9994.000000
mean       28.656896
std       234.260108
min     -6599.978000
25%         1.728750
50%         8.666500
75%        29.364000
max      8399.976000
Name: Profit, dtype: float64


We can notice that the minimum for the profit is in the negatives. This means that this column accounts for the losses as well. 

# Create Meaningful text representations of the data 

In [15]:
#looking again at which columns are present in the dataframe to decide on which ones are important
print(df.columns)

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State',
       'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category',
       'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit'],
      dtype='object')


For each column, the following decision was taken for the textual representation: 
* Row ID : dropped as it does not provide any meaningful information 
* Order ID: dropped at it does not provide meaningful information 
* Order Date: kept for trend analysis 
* Ship Date : kept for trend analysis 
* Ship Mode: kept as it may provide meaningful information 
* Customer ID: dropped as it does not provide meaningful information since we already have customer name
* Customer Name: kept as it may provide meaningful information 
* Segment: kept as it may provide meaningful information 
* Country: dropped at it only includes one unique value (United States)
* City: kept as it provides meaningful information 
* State: kept as it provides meaningful information 
* Postal Code: dropped as it is not needed since we have City, State and Region already 
* Region: kept as it provides meaningful information 
* Product ID: dropped as it is not needed since we have Category, Sub-Category and Product Name already 
* Category: kept as it provides meaningful information 
* Sub-Category: kept as it provides meaningful information 
* Product Name: kept as it provides meaningful information 
* Sales: kept as it provides meaningful information 
* Discount: kept as it may provide meaningful information 
* Profit: kept as it may  provide meaningful information 


For the natural language representation description for each transaction, I chose to make it follow a logical narrative starting with who then when then what then where and finally how much so that it is well formatted. I also chose to make it into a paragraph of multiple sentences rather than one long sentence to make it more clear to use by the RAG system. The discount was turned into a percentage for a more natural way to interpret it and the profit and sales were rounded up to two decimals to get rid of unecessary noise. 

In [16]:
#create natural language representation description for each transaction 
all_natural_representation = []

for i, row in df.iterrows():

  #seperate them into sentences so that we get rid of the \n and the white spaces in between each sentence. 
  sentences = [
     f"This order was made by the customer {row["Customer Name"]} who belongs to the segment {row["Segment"]}.",
     f"The customer ordered on the date {str(row["Order Date"].date())} and was shipped on the date {str(row["Ship Date"].date())} using the shipment method {row["Ship Mode"]}.",
     f"The customer ordered the product {row["Product Name"]} from the sub-category {row["Sub-Category"]} which belongs to the category {row["Category"]}.",
     f"The customer resides in the city of {row["City"]} which belongs to the state {row["State"]} and to the region {row["Region"]}.",
     f"The product generated {round(row["Sales"], 2)}$ in sales and was ordered with a quantity of {row["Quantity"]} using a discount of {row["Discount"]*100}% and generated a profit of {round(row["Profit"], 2)}$." #rounded so that there are not too many numbers after the decimal which might just confuse the LLM
  ]

  #join into one paragraph
  text = " ".join(sentences)

  #include all metadata relevant to the transaction
  dict_result = {
    "text": text, 
    "customer name": row["Customer Name"],
    "segment": row["Segment"], 
    "order date": str(row["Order Date"].date()), 
    "ship date": str(row["Ship Date"].date()), 
    "category": row["Category"],
    "sub-category": row["Sub-Category"],
    "product name": row["Product Name"], 
    "city": row["City"], 
    "state": row["State"], 
    "region": row["Region"], 
    "sales": round(row["Sales"], 2), 
    "quantity": row["Quantity"],
    "discount": row["Discount"], 
    "profit": round(row["Profit"], 2)
  }

  all_natural_representation.append(dict_result)

#save into json as it is suited for text + metadata and easy to access
with open("json_files2/textual_representation.json", "w", encoding="utf-8") as f:
    json.dump(all_natural_representation, f, indent=4, ensure_ascii=False)



In [17]:
#check 
json_df = pd.read_json("json_files2/textual_representation.json")
print(json_df.iloc[0, 0])
print(json_df.shape)

This order was made by the customer Claire Gute who belongs to the segment Consumer. The customer ordered on the date 2016-11-08 and was shipped on the date 2016-11-11 using the shipment method Second Class. The customer ordered the product Bush Somerset Collection Bookcase from the sub-category Bookcases which belongs to the category Furniture. The customer resides in the city of Henderson which belongs to the state Kentucky and to the region South. The product generated 261.96$ in sales and was ordered with a quantity of 2 using a discount of 0.0% and generated a profit of 41.91$.
(9994, 15)


The data was saved into a json file which includes both the text and the metadata. This allows for easier access and providing the metadata allows for the system to retrieve information without going through the whole text and parsing it. This also allows to not have to run the code many times as the information can be directly accessed through the json file. Especially since the dataset contains many rows. 

# Create the aggregated summaries

**For the monthly sales**

On top of the sum of sales, I chose to include the sum of profit, the number of orders and the mean discount applied for each month to allow for further analysis

In [18]:
monthly_df = df.groupby([df["Order Date"].dt.month.rename("Month")]).agg({
    "Sales": "sum", 
    "Profit": "sum", 
    "Order ID": "count", 
    "Discount": "mean"
}).reset_index()

monthly_df.rename(columns={"Sales":"Total Sales", "Profit": "Total Profit", "Order ID": "Number orders", "Discount": "Mean discount"}, inplace=True) #rename columns for easier access
monthly_df.head(3)

,Month,Total Sales,Total Profit,Number orders,Mean discount
0,1,94924.8356,9134.4461,381,0.149921
1,2,59751.2514,10294.6107,300,0.146233
2,3,205005.4888,28594.6872,696,0.155029


In [19]:
all_monthly_info = []

for _, row in monthly_df.iterrows():

    #convert months from digits into the actual month to keep the natural language aspect
    month_mapper = {
        1: "January", 
        2: "February", 
        3: "March", 
        4: "April", 
        5: "May", 
        6: "June", 
        7: "July", 
        8: "August", 
        9: "September", 
        10: "October",
        11: "November", 
        12: "December"
    }

    text = f"For the month {month_mapper[row["Month"]]}, the total sales were of {round(row["Total Sales"],2)}$, the total profit was of {round(row["Total Profit"], 2)}$, the number of orders was {row["Number orders"]} and the mean discount was {round(row["Mean discount"], 2)}.".strip()

    monthly_result = {
        "text": text,
        "month": month_mapper[row["Month"]],
        "total sales": round(row["Total Sales"], 2), 
        "total profit": round(row["Total Profit"], 2),
        "number orders": row["Number orders"],
        "mean discount": round(row["Mean discount"], 2)
    }

    all_monthly_info.append(monthly_result)

with open("json_files2/monthly_sales.json", "w", encoding="utf-8") as f:
    json.dump(all_monthly_info, f, indent=4, ensure_ascii=False)


In [20]:
#check 
json_df = pd.read_json("json_files2/monthly_sales.json")
print(json_df.iloc[0, 0])
print(json_df.shape)

For the month January, the total sales were of 94924.84$, the total profit was of 9134.45$, the number of orders was 381.0 and the mean discount was 0.15.
(12, 6)


In [21]:
#in addition to months, include also years to allow for analysis specific to the years
year_df = df.groupby([df["Order Date"].dt.year.rename("Year")]).agg({
    "Sales": "sum", 
    "Profit": "sum", 
    "Order ID": "count", 
    "Discount": "mean"
}).reset_index()

year_df.rename(columns={"Sales":"Total Sales", "Profit": "Total Profit", "Order ID": "Number orders", "Discount": "Mean discount"}, inplace=True)
year_df.head(3)

,Year,Total Sales,Total Profit,Number orders,Mean discount
0,2014,484247.4981,49543.9741,1993,0.158284
1,2015,470532.5090,61618.6037,2102,0.155609
2,2016,609205.5980,81795.1743,2587,0.154743


In [22]:
all_year_info = []

for _, row in year_df.iterrows():


    text = f"For the year {int(row["Year"])}, the total sales were of {round(row["Total Sales"],2)}$, the total profit was of {round(row["Total Profit"], 2)}$, the number of orders was {row["Number orders"]} and the mean discount was {round(row["Mean discount"], 2)}.".strip()

    year_result = {
        "text": text,
        "year": int(row["Year"]),
        "total sales": round(row["Total Sales"], 2), 
        "total profit": round(row["Total Profit"], 2),
        "number orders": row["Number orders"],
        "mean discount": round(row["Mean discount"], 2)
    }

    all_year_info.append(year_result)

with open("json_files2/year_sales.json", "w", encoding="utf-8") as f:
    json.dump(all_year_info, f, indent=4, ensure_ascii=False)


In [23]:
#check 
json_df = pd.read_json("json_files2/year_sales.json")
print(json_df.iloc[0, 0])
print(json_df.shape)

For the year 2014, the total sales were of 484247.5$, the total profit was of 49543.97$, the number of orders was 1993.0 and the mean discount was 0.16.
(4, 6)


**For the category performance** 

In [24]:
category_df = df.groupby([df["Category"]]).agg({
    "Sales": "sum", 
    "Profit": "sum", 
    "Order ID": "count", 
    "Discount": "mean"
}).reset_index()

category_df.rename(columns={"Sales":"Total Sales", "Profit": "Total Profit", "Order ID": "Number orders", "Discount": "Mean discount"}, inplace=True)
category_df.head(5)

,Category,Total Sales,Total Profit,Number orders,Mean discount
0,Furniture,741999.7953,18451.2728,2121,0.173923
1,Office Supplies,719047.0320,122490.8008,6026,0.157285
2,Technology,836154.0330,145454.9481,1847,0.132323


In [25]:
all_category_info = []

for i, row in category_df.iterrows(): 
    text = f"For the category {row["Category"]}, the total sales were of {round(row["Total Sales"],2)}$, the total profit was of {round(row["Total Profit"], 2)}$, the number of orders was {row["Number orders"]} and the mean discount was {round(row["Mean discount"], 2)}.".strip()

    category_result = {
        "text": text,
        "category": row["Category"],
        "total sales": round(row["Total Sales"], 2), 
        "total profit": round(row["Total Profit"], 2),
        "number orders": row["Number orders"],
        "mean discount": round(row["Mean discount"], 2)
    }

    all_category_info.append(category_result)

with open("json_files2/category_info.json", "w", encoding="utf-8") as f:
    json.dump(all_category_info, f, indent=4, ensure_ascii=False)



In [26]:
#check 
json_df = pd.read_json("json_files2/category_info.json")
print(json_df.iloc[0, 0])
print(json_df.shape)

For the category Furniture, the total sales were of 741999.8$, the total profit was of 18451.27$, the number of orders was 2121 and the mean discount was 0.17.
(3, 6)


In [27]:
#In addition to category, subcategory allows for a layer more in depth and can be used when we want to analyze the categories by what constitutes them
subcategory_df = df.groupby([df["Sub-Category"]]).agg({
    "Sales": "sum", 
    "Profit": "sum", 
    "Order ID": "count", 
    "Discount": "mean"
}).reset_index()

subcategory_df.rename(columns={"Sales":"Total Sales", "Profit": "Total Profit", "Order ID": "Number orders", "Discount": "Mean discount"}, inplace=True)
subcategory_df.head(5)

,Sub-Category,Total Sales,Total Profit,Number orders,Mean discount
0,Accessories,167380.3180,41936.6357,775,0.078452
1,Appliances,107532.1610,18138.0054,466,0.166524
2,Art,27118.7920,6527.7870,796,0.074874
3,Binders,203412.7330,30221.7633,1523,0.372292
4,Bookcases,114879.9963,-3472.5560,228,0.211140


In [28]:
all_subcategory_info = []

for i, row in subcategory_df.iterrows(): 
    text = f"For the sub-category {row["Sub-Category"]}, the total sales were of {round(row["Total Sales"],2)}$, the total profit was of {round(row["Total Profit"], 2)}$, the number of orders was {row["Number orders"]} and the mean discount was {round(row["Mean discount"], 2)}.".strip()

    subcategory_result = {
        "text": text,
        "category": row["Sub-Category"],
        "total sales": round(row["Total Sales"], 2), 
        "total profit": round(row["Total Profit"], 2),
        "number orders": row["Number orders"],
        "mean discount": round(row["Mean discount"], 2)
    }

    all_subcategory_info.append(subcategory_result)

with open("json_files2/subcategory_info.json", "w", encoding="utf-8") as f:
    json.dump(all_subcategory_info, f, indent=4, ensure_ascii=False)

In [29]:
#check 
json_df = pd.read_json("json_files2/subcategory_info.json")
print(json_df.iloc[0, 0])
print(json_df.shape)

For the sub-category Accessories, the total sales were of 167380.32$, the total profit was of 41936.64$, the number of orders was 775 and the mean discount was 0.08.
(17, 6)


**For the regional analysis**

In [30]:
region_df = df.groupby([df["Region"]]).agg({
    "Sales": "sum", 
    "Profit": "sum", 
    "Order ID": "count", 
    "Discount": "mean"
}).reset_index()

region_df.rename(columns={"Sales":"Total Sales", "Profit": "Total Profit", "Order ID": "Number orders", "Discount": "Mean discount"}, inplace=True)
region_df.head(5)

,Region,Total Sales,Total Profit,Number orders,Mean discount
0,Central,501239.8908,39706.3625,2323,0.240353
1,East,678781.2400,91522.7800,2848,0.145365
2,South,391721.9050,46749.4303,1620,0.147253
3,West,725457.8245,108418.4489,3203,0.109335


In [31]:
all_region_info = []

for i, row in region_df.iterrows():
    text = f"For the region {row["Region"]}, the total sales were of {round(row["Total Sales"],2)}$, the total profit was of {round(row["Total Profit"], 2)}$, the number of orders was {row["Number orders"]} and the mean discount was {round(row["Mean discount"], 2)}.".strip()

    region_result = {
        "text": text,
        "region": row["Region"],
        "total sales": round(row["Total Sales"], 2), 
        "total profit": round(row["Total Profit"], 2),
        "number orders": row["Number orders"],
        "mean discount": round(row["Mean discount"], 2)
    }

    all_region_info.append(region_result)

with open("json_files2/region_info.json", "w", encoding="utf-8") as f:
    json.dump(all_region_info, f, indent=4, ensure_ascii=False)


    

In [32]:
#check 
json_df = pd.read_json("json_files2/region_info.json")
print(json_df.iloc[0, 0])
print(json_df.shape)

For the region Central, the total sales were of 501239.89$, the total profit was of 39706.36$, the number of orders was 2323 and the mean discount was 0.24.
(4, 6)


In [33]:
#in addition to region data which might too large, we have state and city information which allows for more specific analysis. These were added to allow for quick access when wanting specifically state or city data
state_df = df.groupby([df["State"]]).agg({
    "Sales": "sum", 
    "Profit": "sum", 
    "Order ID": "count", 
    "Discount": "mean"
}).reset_index()

state_df.rename(columns={"Sales":"Total Sales", "Profit": "Total Profit", "Order ID": "Number orders", "Discount": "Mean discount"}, inplace=True)
state_df.head(5)

,State,Total Sales,Total Profit,Number orders,Mean discount
0,Alabama,19510.6400,5786.8253,61,0.000000
1,Arizona,35282.0010,-3427.9246,224,0.303571
2,Arkansas,11678.1300,4008.6871,60,0.000000
3,California,457687.6315,76381.3871,2001,0.072764
4,Colorado,32108.1180,-6527.8579,182,0.316484


In [34]:
all_state_info = []

for i, row in state_df.iterrows():
    text = f"For the state {row["State"]}, the total sales were of {round(row["Total Sales"],2)}$, the total profit was of {round(row["Total Profit"], 2)}$, the number of orders was {row["Number orders"]} and the mean discount was {round(row["Mean discount"], 2)}.".strip()

    state_result = {
        "text": text,
        "state": row["State"],
        "total sales": round(row["Total Sales"], 2), 
        "total profit": round(row["Total Profit"], 2),
        "number orders": row["Number orders"],
        "mean discount": round(row["Mean discount"], 2)
    }

    all_state_info.append(state_result)

with open("json_files2/state_info.json", "w", encoding="utf-8") as f:
    json.dump(all_state_info, f, indent=4, ensure_ascii=False)

In [35]:
#check 
json_df = pd.read_json("json_files2/state_info.json")
print(json_df.iloc[0, 0])
print(json_df.shape)

For the state Alabama, the total sales were of 19510.64$, the total profit was of 5786.83$, the number of orders was 61 and the mean discount was 0.0.
(49, 6)


In [36]:
city_df = df.groupby([df["City"]]).agg({
    "Sales": "sum", 
    "Profit": "sum", 
    "Order ID": "count", 
    "Discount": "mean"
}).reset_index()

city_df.rename(columns={"Sales":"Total Sales", "Profit": "Total Profit", "Order ID": "Number orders", "Discount": "Mean discount"}, inplace=True)
city_df.head(5)

,City,Total Sales,Total Profit,Number orders,Mean discount
0,Aberdeen,25.500,6.6300,1,0.000000
1,Abilene,1.392,-3.7584,1,0.800000
2,Akron,2729.986,-186.6356,21,0.357143
3,Albuquerque,2220.160,634.0881,14,0.057143
4,Alexandria,5519.570,318.6183,16,0.000000


In [37]:
all_city_info = []

for i, row in city_df.iterrows():
    text = f"For the city {row["City"]}, the total sales were of {round(row["Total Sales"],2)}$, the total profit was of {round(row["Total Profit"], 2)}$, the number of orders was {row["Number orders"]} and the mean discount was {round(row["Mean discount"], 2)}.".strip()

    city_result = {
        "text": text,
        "city": row["City"],
        "total sales": round(row["Total Sales"], 2), 
        "total profit": round(row["Total Profit"], 2),
        "number orders": row["Number orders"],
        "mean discount": round(row["Mean discount"], 2)
    }

    all_city_info.append(city_result)

with open("json_files2/city_info.json", "w", encoding="utf-8") as f:
    json.dump(all_city_info, f, indent=4, ensure_ascii=False)

In [38]:
#check 
json_df = pd.read_json("json_files2/city_info.json")
print(json_df.iloc[0, 0])
print(json_df.shape)

For the city Aberdeen, the total sales were of 25.5$, the total profit was of 6.63$, the number of orders was 1 and the mean discount was 0.0.
(531, 6)


# Statistical summaries 

In [39]:
#global summaries, allow for a global overview of the data and to get general answers
total_sales = df["Sales"].sum()
print("The total sales are: ", total_sales)
total_profit = df["Profit"].sum()
print("The total profit is: ", total_profit)
total_num_orders = df["Order ID"].count()
print("The total number of orders are: ", total_num_orders)
total_average_discount = df["Discount"].mean()
print("The average discount is: ", total_average_discount)
margin_of_profit = total_profit / total_sales
print("The margin of profit is : ", margin_of_profit)
mean_sale_order = total_sales / total_num_orders
print("The mean sales per order is : ", mean_sale_order)



The total sales are:  2297200.8603000003
The total profit is:  286397.0217
The total number of orders are:  9994
The average discount is:  0.15620272163297977
The margin of profit is :  0.12467217240315603
The mean sales per order is :  229.85800083049833


In [40]:
setences = [
    f"The total sales for the dataset is {round(total_sales,2)}$.",
    f"The total profit for the dataset is {round(total_profit, 2)}$.",
    f"The total number of orders for the dataset is {total_num_orders}.",
    f"The average discount for the dataset is {round(total_average_discount, 2)}.",
    f"The margin of profit for the dataset is {round(margin_of_profit, 2)}.",
    f"The mean sale per order for the dataset is {round(mean_sale_order, 2)}$."
]

global_info = " ".join(setences)
print(global_info)

general_dict = {
    "text": global_info,
    "total sales": total_sales, 
    "total profit": total_profit,
    "total number orders": total_num_orders,
    "average discount": total_average_discount,
    "margin of profit": margin_of_profit,
    "mean sale per order": mean_sale_order
}

The total sales for the dataset is 2297200.86$. The total profit for the dataset is 286397.02$. The total number of orders for the dataset is 9994. The average discount for the dataset is 0.16. The margin of profit for the dataset is 0.12. The mean sale per order for the dataset is 229.86$.


In [41]:
#in addition to the general information, it is also important to get some summaries about time data as well 
first_date = df["Order Date"].dt.date.min()
print("The first order date is: ", first_date)
last_date = df["Order Date"].dt.date.max()
print("The last order date is: ", last_date)
time_span = last_date.year - first_date.year
print("The covered time span is: ", time_span)
df["Shipping Time"] = df["Ship Date"] - df["Order Date"]
df["Shipping Time Days"] = df["Shipping Time"].dt.days
average_shipping = df["Shipping Time Days"].mean()
print("The average time it takes for shipping is: ", average_shipping)

The first order date is:  2014-01-03
The last order date is:  2017-12-30
The covered time span is:  3
The average time it takes for shipping is:  3.958174904942966


In [42]:
sentences = [
    f"The first order date is {first_date}.", 
    f"The last order date is {last_date}.", 
    f"The covered time span is {time_span} years.",
    f"The average shipping time is {round(average_shipping,2)} days."
]

time_info = " ".join(sentences)
print(time_info)

time_dict = {
    "text": time_info,
    "first order": first_date,
    "last order": last_date,
    "time span": time_span,
    "average shipping time": average_shipping
}

The first order date is 2014-01-03. The last order date is 2017-12-30. The covered time span is 3 years. The average shipping time is 3.96 days.


In [43]:
#both global and time info were merged into one and saved into a dictionary. To seperate the two, additional labels are assigned to each accordingly 
all_general_text = [global_info, time_info]
labels = ["global", "time"]

final_result = []

for i, text in enumerate(all_general_text): 
    global_dict = {
        "text": text,
        "label": labels[i]
    }

    final_result.append(global_dict)

with open("json_files2/stats_summary.json", "w", encoding="utf-8") as f:
    json.dump(final_result, f, indent=4, ensure_ascii=False)


In [44]:
json_df = pd.read_json("json_files2/stats_summary.json")
json_df

,text,label
0,The total sales for the dataset is 2297200.86$...,global
1,The first order date is 2014-01-03. The last o...,time


# Chunking strategies 

In [45]:
# combine everything into one part 
final_files = [
    "json_files2/textual_representation.json",
    "json_files2/category_info.json",
    "json_files2/city_info.json",
    "json_files2/monthly_sales.json",
    "json_files2/region_info.json",
    "json_files2/state_info.json",
    "json_files2/stats_summary.json",
    "json_files2/subcategory_info.json",
    "json_files2/year_sales.json"
]

combined_data = []
#useee different type files labels accordingly in order to seperate between each type of document, also allows for easier filtering in later steps
types_files = ["transaction", "category_summary", "city_summary", "month_summary", "region_summary", "state_summary", "general_summary", "subcategory_summary", "year_summary"]

for i, file in enumerate(final_files):  
    with open(file, "r", encoding="utf-8") as f:
        data = json.load(f)

    for d in data: 
        d["type"] = types_files[i]

    combined_data.extend(data)



In [46]:
len(combined_data)

10616

In [47]:
with open("json_files2/combined.json", "w", encoding="utf-8") as f:
    json.dump(combined_data, f, indent=4, ensure_ascii=False)

In [48]:
#chunk based on results
def chunking(text, size): 
    chunks = []
    for i in range(0, len(text), size):
        chunks.append(text[i:i+size])

    return chunks 

In [49]:
#gather everything into one place 
def chunk_results(data, size):
    chunk_results = []

    for d in data: 
        chunks = chunking(d["text"], size)

        for i, chunk in enumerate(chunks):
            chunk_df = d.copy()
            chunk_df["id"] = i #add a unique identifier to be used later on in chromaDB
            chunk_df["text"] = chunk
            chunk_results.append(chunk_df)

    return chunk_results      
        

In [50]:
#check document length information to help decide on most suited chunking strategies
length_document = [len(c["text"]) for c in combined_data]
print(f"Minimum number characters: {min(length_document)}, Maximum number characters: {max(length_document)}, Average number characters: {sum(length_document)/len(length_document):.0f}")


Minimum number characters: 138, Maximum number characters: 690, Average number characters: 572


In [51]:
result_500 = chunk_results(combined_data, 500)
result_500

[{'text': 'This order was made by the customer Claire Gute who belongs to the segment Consumer. The customer ordered on the date 2016-11-08 and was shipped on the date 2016-11-11 using the shipment method Second Class. The customer ordered the product Bush Somerset Collection Bookcase from the sub-category Bookcases which belongs to the category Furniture. The customer resides in the city of Henderson which belongs to the state Kentucky and to the region South. The product generated 261.96$ in sales and wa',
  'customer name': 'Claire Gute',
  'segment': 'Consumer',
  'order date': '2016-11-08',
  'ship date': '2016-11-11',
  'category': 'Furniture',
  'sub-category': 'Bookcases',
  'product name': 'Bush Somerset Collection Bookcase',
  'city': 'Henderson',
  'state': 'Kentucky',
  'region': 'South',
  'sales': 261.96,
  'quantity': 2,
  'discount': 0.0,
  'profit': 41.91,
  'type': 'transaction',
  'id': 0},
 {'text': 's ordered with a quantity of 2 using a discount of 0.0% and genera

It can be noticed here that since 500 is smaller than the average number of characters which is 572, the text gets split weirdly in the middle, making the fixed-size 500 characters approach not suitable 

In [52]:
result_1000 = chunk_results(combined_data, 1000)
result_1000

[{'text': 'This order was made by the customer Claire Gute who belongs to the segment Consumer. The customer ordered on the date 2016-11-08 and was shipped on the date 2016-11-11 using the shipment method Second Class. The customer ordered the product Bush Somerset Collection Bookcase from the sub-category Bookcases which belongs to the category Furniture. The customer resides in the city of Henderson which belongs to the state Kentucky and to the region South. The product generated 261.96$ in sales and was ordered with a quantity of 2 using a discount of 0.0% and generated a profit of 41.91$.',
  'customer name': 'Claire Gute',
  'segment': 'Consumer',
  'order date': '2016-11-08',
  'ship date': '2016-11-11',
  'category': 'Furniture',
  'sub-category': 'Bookcases',
  'product name': 'Bush Somerset Collection Bookcase',
  'city': 'Henderson',
  'state': 'Kentucky',
  'region': 'South',
  'sales': 261.96,
  'quantity': 2,
  'discount': 0.0,
  'profit': 41.91,
  'type': 'transaction',


Here since 1000 characters is just slightly above the maximum number of characters, none of the documents get split in the middle and not too much noise is added while also allowing for some flexibility, making it therefore a good potential final chunking strategy option.

In [53]:
result_2000 = chunk_results(combined_data, 2000)
result_2000

[{'text': 'This order was made by the customer Claire Gute who belongs to the segment Consumer. The customer ordered on the date 2016-11-08 and was shipped on the date 2016-11-11 using the shipment method Second Class. The customer ordered the product Bush Somerset Collection Bookcase from the sub-category Bookcases which belongs to the category Furniture. The customer resides in the city of Henderson which belongs to the state Kentucky and to the region South. The product generated 261.96$ in sales and was ordered with a quantity of 2 using a discount of 0.0% and generated a profit of 41.91$.',
  'customer name': 'Claire Gute',
  'segment': 'Consumer',
  'order date': '2016-11-08',
  'ship date': '2016-11-11',
  'category': 'Furniture',
  'sub-category': 'Bookcases',
  'product name': 'Bush Somerset Collection Bookcase',
  'city': 'Henderson',
  'state': 'Kentucky',
  'region': 'South',
  'sales': 261.96,
  'quantity': 2,
  'discount': 0.0,
  'profit': 41.91,
  'type': 'transaction',


It can be noticed that the 2000 characters approach gives the exact same results as the 1000 characters approach while also adding more unecessary noise, making it therefore not needed.

In [54]:
#save them 
with open("json_files2/chunks_500.json", "w", encoding="utf-8") as f:
    json.dump(result_500, f, indent=4, ensure_ascii=False)

with open("json_files2/chunks_1000.json", "w", encoding="utf-8") as f:
    json.dump(result_1000, f, indent=4, ensure_ascii=False)

with open("json_files2/chunks_2000.json", "w", encoding="utf-8") as f:
    json.dump(result_2000, f, indent=4, ensure_ascii=False)

It can be noticed that with a size of 500, the sentences get cut in the middle. With 1000 it seems to capture the full thing. The result for 2000 returns the same as of 1000 so the additional 1000 are unecessary. Therefore the best solution is 1000. This is also due to the number of average characters we have of 573 and the maximum number of 690. 

# Document-aware Chunking approach

In [55]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

C:\Users\USER\AppData\Roaming\Python\Python313\site-packages\requests\__init__.py:113: RequestsDependencyWarning: urllib3 (2.5.0) or chardet (7.0.1)/charset_normalizer (3.4.3) doesn't match a supported version!
  warnings.warn(


In [56]:
#experimenting with different chunk_sizes and chunk_overlap https://lagnchain.readthedocs.io/en/stable/modules/indexes/text_splitters/examples/recursive_text_splitter.html
def split_document(data, chunk_size, chunk_overlap):
    splitter = RecursiveCharacterTextSplitter(
    chunk_size=chunk_size,
    chunk_overlap=chunk_overlap, #allows for some overlap between the chunks to allow for the context to not be cut suddenly
    separators=["\n\n", "\n", ". ", " ", ""] #order of splitting first paragraphs then sentences etc 
    )

    chunked = []
    for d in data: 
        chunks = splitter.split_text(d["text"])

        for i, chunk in enumerate(chunks):
            chunk_df = d.copy()
            chunk_df["id"] = i
            chunk_df["text"] = chunk
            chunked.append(chunk_df)
    return chunked




In [57]:
split_document(combined_data, 500, 100)

[{'text': 'This order was made by the customer Claire Gute who belongs to the segment Consumer. The customer ordered on the date 2016-11-08 and was shipped on the date 2016-11-11 using the shipment method Second Class. The customer ordered the product Bush Somerset Collection Bookcase from the sub-category Bookcases which belongs to the category Furniture. The customer resides in the city of Henderson which belongs to the state Kentucky and to the region South',
  'customer name': 'Claire Gute',
  'segment': 'Consumer',
  'order date': '2016-11-08',
  'ship date': '2016-11-11',
  'category': 'Furniture',
  'sub-category': 'Bookcases',
  'product name': 'Bush Somerset Collection Bookcase',
  'city': 'Henderson',
  'state': 'Kentucky',
  'region': 'South',
  'sales': 261.96,
  'quantity': 2,
  'discount': 0.0,
  'profit': 41.91,
  'type': 'transaction',
  'id': 0},
 {'text': '. The product generated 261.96$ in sales and was ordered with a quantity of 2 using a discount of 0.0% and genera

For 500 characters, this method demonstrated better results than fixed chunking as the sentence is no longer cut in the middle but cut in a natural boundary. However the full text is still not included. 

In [58]:
split_document(combined_data, 1000, 100)

[{'text': 'This order was made by the customer Claire Gute who belongs to the segment Consumer. The customer ordered on the date 2016-11-08 and was shipped on the date 2016-11-11 using the shipment method Second Class. The customer ordered the product Bush Somerset Collection Bookcase from the sub-category Bookcases which belongs to the category Furniture. The customer resides in the city of Henderson which belongs to the state Kentucky and to the region South. The product generated 261.96$ in sales and was ordered with a quantity of 2 using a discount of 0.0% and generated a profit of 41.91$.',
  'customer name': 'Claire Gute',
  'segment': 'Consumer',
  'order date': '2016-11-08',
  'ship date': '2016-11-11',
  'category': 'Furniture',
  'sub-category': 'Bookcases',
  'product name': 'Bush Somerset Collection Bookcase',
  'city': 'Henderson',
  'state': 'Kentucky',
  'region': 'South',
  'sales': 261.96,
  'quantity': 2,
  'discount': 0.0,
  'profit': 41.91,
  'type': 'transaction',


In [59]:
split_document(combined_data, 2000, 100)

[{'text': 'This order was made by the customer Claire Gute who belongs to the segment Consumer. The customer ordered on the date 2016-11-08 and was shipped on the date 2016-11-11 using the shipment method Second Class. The customer ordered the product Bush Somerset Collection Bookcase from the sub-category Bookcases which belongs to the category Furniture. The customer resides in the city of Henderson which belongs to the state Kentucky and to the region South. The product generated 261.96$ in sales and was ordered with a quantity of 2 using a discount of 0.0% and generated a profit of 41.91$.',
  'customer name': 'Claire Gute',
  'segment': 'Consumer',
  'order date': '2016-11-08',
  'ship date': '2016-11-11',
  'category': 'Furniture',
  'sub-category': 'Bookcases',
  'product name': 'Bush Somerset Collection Bookcase',
  'city': 'Henderson',
  'state': 'Kentucky',
  'region': 'South',
  'sales': 261.96,
  'quantity': 2,
  'discount': 0.0,
  'profit': 41.91,
  'type': 'transaction',


While document specific chunking could be a better solution for larger documents, since my documents have an average number of characters of 573, the results for 1000 characters and 2000 characters are the same as obtained with fixed chunking and therefore a fixed-chunking with 1000 characters remains the most ideal and straight-forward solution for this case. 